client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx

```diff
@@ -0,0 +1,133 @@
+import FormSettingExplanation from 'calypso/components/forms/form-setting-explanation';
+import InlineSupportLink from 'calypso/components/inline-support-link';
+import type { DomainData, SiteDetails } from '@automattic/data-stores';
+
+import './style.scss';
+
+type PrimaryDomainSelectorProps = {
+	domains: undefined | DomainData[];
+	userCanSetPrimaryDomains: boolean;
+	site: undefined | null | SiteDetails;
+	onSetPrimaryDomain: ( domain: string, onComplete: () => void, type: string ) => void;
+};
+
+const PrimaryDomainSelector = ( {
+	domains,
+	site,
+	userCanSetPrimaryDomains,
+	onSetPrimaryDomain,
+}: PrimaryDomainSelectorProps ) => {
+	const [ selectedDomain, setSelectedDomain ] = useState< undefined | string >( undefined );
+	const [ isSettingPrimaryDomain, setIsSettingPrimaryDomain ] = useState< boolean >( false );
+
+	const translate = useTranslate();
+
+	useEffect( () => {
+		if ( domains?.length ) {
+			const primaryDomain = domains.find( ( domain ) => {
+				return domain.primary_domain;
+			} );
+			if ( primaryDomain ) {
+				setSelectedDomain( primaryDomain.domain );
+			}
+		}
+	}, [ domains ] );
+
+	if ( ! domains || ! site ) {
+		return null;
+	}
+
+	const isOnFreePlan = site?.plan?.is_free ?? false;
+
+	const shouldUpgradeToMakeDomainPrimary = ( domain: DomainData ): boolean => {
+		return (
+			! userCanSetPrimaryDomains &&
+			isOnFreePlan &&
+			( domain.type === 'registered' || domain.type === 'mapping' ) &&
+			! domain.current_user_can_create_site_from_domain_only &&
+			! domain.wpcom_domain &&
+			! domain.is_wpcom_staging_domain &&
+			( site?.plan?.features.active.includes( FEATURE_SET_PRIMARY_CUSTOM_DOMAIN ) ?? false )
+		);
+	};
+
+	let validPrimaryDomains = domains.filter( ( domain ) => {
+		return (
+			domain.can_set_as_primary &&
+			! domain.aftermarket_auction &&
+			domain.points_to_wpcom &&
+			! shouldUpgradeToMakeDomainPrimary( domain )
+		);
+	} );
+
+	const hasWpcomStagingDomain = validPrimaryDomains.find(
+		( domain ) => domain.is_wpcom_staging_domain
+	);
+
+	if ( hasWpcomStagingDomain ) {
+		validPrimaryDomains = validPrimaryDomains.filter( ( domain ) => {
+			if ( domain.wpcom_domain ) {
+				return domain.is_wpcom_staging_domain;
+			}
+
+			return true;
+		} );
+	}
+
+	const onSelectChange = ( event: ChangeEvent< HTMLSelectElement > ) => {
+		setSelectedDomain( event.target.value );
+
+		const domain = domains.find( ( d ) => d.domain === event.target.value );
+		if ( ! domain ) {
+			return;
+		}
+
+		setIsSettingPrimaryDomain( true );
+		onSetPrimaryDomain( event.target.value, () => setIsSettingPrimaryDomain( false ), domain.type );
+	};
+
+	const supportLink = localizeUrl( 'https://wordpress.com/support/domains/set-a-primary-address/' );
+
+	return (
+		<FormFieldset className="domains-set-primary-address">
+			<FormLabel htmlFor="primary-domain-selector" className="domains-set-primary-address__title">
+				{ translate( 'Primary site address' ) }
```

Comment:
What do you think of making this an action verb so that it's a little clearer? (Not a leading question, I'm not sure about this!)

```suggestion
				{ translate( 'Change primary site address' ) }
```

If you wanted to be funky with the language here, you could even give an indication of what the primary site address refers to without needing another click. 

```suggestion
				{ translate( 'Looking to change the site address used by your viewers?' ) }
```